fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java

```diff
@@ -24,8 +25,35 @@ protected void enumerate() {
             return (diff < 0 ? -1 : (diff > 0 ? 1 : 0));
         });
 
+        boolean useHeuristic = true;
+        boolean[] used = new boolean[atomSize];
         GroupInfo leftGroup = atoms.get(0);
-        for (int index = 1; index < atomSize; ++index) {
+        used[0] = true;
+        int next = 1;
+        while (next < atomSize) {
+            if (used[next]) {
+                next++;
+                continue;
+            }
+            int index = next;
+            Preconditions.checkState(!used[index]);
```

Comment:
Seems this check is unnecessary

Reply:
defensive programming:)

Reply:
Emm, that's not defensive, it is sure that used[index] must be false